Foundation/System/Bindings/POSIX/Stropts.hsc

```diff
@@ -0,0 +1,235 @@
+        #{poke ConsoleScreenBufferInfo, dwSize} ptr s
+        #{poke ConsoleScreenBufferInfo, dwCursorPosition} ptr c
+        #{poke ConsoleScreenBufferInfo, wAttributes} ptr a
+        #{poke ConsoleScreenBufferInfo, srWindow} ptr w
+        #{poke ConsoleScreenBufferInfo, dwMaximumWindowSize} ptr m
+    
+invalidHandleValue :: IntPtr
+invalidHandleValue = #{const INVALID_HANDLE_VALUE}
+
+stdInputHandle :: CULong
+stdInputHandle = #{const STD_INPUT_HANDLE}
+#endif
+-- defined FOUNDATION_SYSTEM_WINDOWS
+
+
+foreign import ccall "popen" c_popen :: CString -> CString -> IO (Ptr CFile)
+foreign import ccall "pclose" c_pclose :: Ptr CFile -> IO CInt
+foreign import ccall "fscanf_int" c_fscanf_int :: Ptr CFile -> Ptr CInt -> IO CInt
+#ifdef FOUNDATION_SYSTEM_UNIX
+foreign import ccall "ioctl_winsize" c_ioctl_winsize :: CInt -> Ptr Winsize -> IO CInt
+#elif defined FOUNDATION_SYSTEM_WINDOWS
+foreign import ccall "GetStdHandle" c_get_std_handle :: CULong -> IO Handle
+foreign import ccall "GetConsoleScreenBufferInfo" c_get_console_screen_buffer_info :: Handle -> Ptr ConsoleScreenBufferInfo -> IO CInt
+#endif
+
+#ifdef FOUNDATION_SYSTEM_UNIX
+ioctlWinsize :: CInt -> IO (Maybe (Int, Int))
+ioctlWinsize fd = do
+    winsizePtr <- malloc :: IO (Ptr Winsize)
+    status <- c_ioctl_winsize fd winsizePtr
+    winsize <- peek winsizePtr
+    free winsizePtr
+    if' (fromIntegral status == (-1 :: Int))
+       (return Nothing)
+       (return $ Just (fromIntegral . ws_col $ winsize, fromIntegral . ws_row $ winsize))
+       
+#elif defined FOUNDATION_SYSTEM_WINDOWS
+getStdHandle :: CULong -> IO (Maybe Handle)
+getStdHandle handleRef = do
+    handle <- c_get_std_handle handleRef
+    if' (handle == intPtrToPtr invalidHandleValue)
+       (return Nothing)
+       (return Just handle)
+
+getConsoleScreenBufferInfo :: Handle -> IO (Maybe ConsoleScreenBufferInfo)
+getConsoleScreenBufferInfo handle = do
+    infoPtr <- malloc :: IO (Ptr ConsoleScreenBufferInfo)
+    status <- c_get_console_screen_buffer_info handle infoPtr
+    info <- peek infoPtr
+    free infoPtr
+    if' (status == 0)
+       (return Nothing)
+       (return Just info)
+       
+winWinsize :: CULong -> IO (Maybe (Int, Int))
+winWinsize handleRef = do
+    maybeHandle <- getStdHandle handleRef
+    case maybeHandle of
+        Nothing -> return Nothing
+        Just handle -> do
+            maybeInfo <- getConsoleScreenBufferInfo handle
+            case maybeInfo of
+                Nothing -> return Nothing
+                Just info -> do
+                    let window = srWindow info
+                        width = (right window - left window + 1)
+                        height = (bottom window - top window + 1)
+                    return (width height)
+#endif
+-- defined FOUNDATION_SYSTEM_WINDOWS
+            
+if' :: Bool -> a -> a -> a
+if' True t _ = t
+if' False _ f = f
+
+popen :: String -> String -> IO (Ptr CFile)
+popen cmd_str mode_str = do
```

Comment:
I don't think we need any popen thing; Personally I'm happy if we just return a constant (80x25) when the first method of gathering information fails. `popen` doesn't really play nice with the runtime and it's doing a lots of work (fork, exec, setting up pipe, read/write) which are blocking and assume lots of things